File: apple-client/Hearth/Core/Sources/HearthCore/Config/ServerConfig.swift
//
//  ServerConfig.swift
//  Hearth
//
//  Where the house is, and the one type that builds an origin.
//
//  The parser below carries across from Valinor unchanged, because it was paid
//  for in bugs. What changed is the default: there is none.
//

import Foundation

public final class ServerConfig {
    public static let shared = ServerConfig()

    private let serverHostKey = "hearth.serverHost"
    private let serverPortKey = "hearth.serverPort"
    private let deviceTokenKey = "hearth.deviceToken"

    /// 18700 is Hearth's client gateway, and the port is the whole default.
    ///
    /// NOT 8700. The development machine runs Valinor's stack there, so a
    /// Hearth build defaulting to 8700 does not fail -- it connects, and comes
    /// up wearing someone else's memory, journal and personas. That is a first
    /// run which looks flawless and is the worst outcome available.
    public static let defaultPort = 18700

    /// **There is deliberately no default host.**
    ///
    /// The desktop client defaults to 127.0.0.1 because it supervises its own
    /// backend and the house genuinely is on that machine. A phone supervises
    /// nothing and is never the machine running the backend, so neither
    /// available default is honest: 127.0.0.1 on a phone is the phone, and a
    /// LAN literal is one machine on one home network compiled into a product.
    ///
    /// Unset is therefore a distinct state rather than a missing value, and the
    /// app in that state does not dial. It renders the bundled persona and asks
    /// where the house is. That makes first run correct by construction instead
    /// of correct by a timeout expiring.
    ///
    /// Bonjour discovery is the obvious later refinement and does not need to
    /// exist for this to be right.
    public var serverHost: String? {
        get {
            guard let saved = UserDefaults.standard.string(forKey: serverHostKey),
                  !saved.isEmpty else { return nil }
            return saved
        }
        set {
            let trimmed = newValue?.trimmingCharacters(in: .whitespacesAndNewlines)
            if let trimmed, !trimmed.isEmpty {
                UserDefaults.standard.set(trimmed, forKey: serverHostKey)
            } else {
                UserDefaults.standard.removeObject(forKey: serverHostKey)
            }
        }
    }

    public var serverPort: Int {
        get {
            let saved = UserDefaults.standard.integer(forKey: serverPortKey)
            return saved > 0 ? saved : Self.defaultPort
        }
        set { UserDefaults.standard.set(newValue, forKey: serverPortKey) }
    }

    /// The device token this phone was given when it paired, or nil.
    ///
    /// A house exempts loopback and requires a token from everything else, so
    /// a phone always needs one -- a phone is never the machine running the
    /// backend. It is stored beside the address rather than in the Keychain
    /// for now, which is a deliberate pre-alpha shortcut and not a decision:
    /// UserDefaults is readable from a backup, and the Keychain is where this
    /// belongs before anyone who is not us installs it.
    public var deviceToken: String? {
        get {
            guard let saved = UserDefaults.standard.string(forKey: deviceTokenKey),
                  !saved.isEmpty else { return nil }
            return saved
        }
        set {
            let trimmed = newValue?.trimmingCharacters(in: .whitespacesAndNewlines)
            if let trimmed, !trimmed.isEmpty {
                UserDefaults.standard.set(trimmed, forKey: deviceTokenKey)
            } else {
                UserDefaults.standard.removeObject(forKey: deviceTokenKey)
            }
        }
    }

    /// True when someone has told this client where the house is. The socket
    /// must not dial while this is false.
    public var isConfigured: Bool { serverHost != nil }

    /// True when this phone has both an address AND a way in. The distinction
    /// matters to first run: knowing where the house is and being allowed
    /// through its door are two different states, and a client that conflates
    /// them shows "connecting" forever instead of "you need to pair".
    public var isPaired: Bool { deviceToken != nil }

    /// What the Connection field shows and accepts: `host` or `host:port`.
    ///
    /// Clearing it clears the host, and that is the change from Valinor, where
    /// clearing restored the build-time default and a stranger's install could
    /// not get away from one person's LAN address.
    public var address: String {
        get {
            guard let host = serverHost else { return "" }
            return serverPort == Self.defaultPort ? host : "\(host):\(serverPort)"
        }
        set {
            let trimmed = newValue.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty else {
                serverHost = nil
                serverPort = Self.defaultPort
                NotificationCenter.default.post(name: .hearthServerConfigured, object: nil)
                return
            }
            defer { NotificationCenter.default.post(name: .hearthServerConfigured, object: nil) }
            // Strip a pasted scheme; people paste URLs.
            var value = trimmed
            for scheme in ["ws://", "wss://", "http://", "https://"] where value.hasPrefix(scheme) {
                value = String(value.dropFirst(scheme.count))
            }
            value = value.trimmingCharacters(in: CharacterSet(charactersIn: "/"))

            // Rightmost colon only, and only when what follows is a port --
            // an IPv6 literal is full of colons and must not be split.
            if let colon = value.lastIndex(of: ":"),
               let port = Int(value[value.index(after: colon)...]),
               port > 0, port < 65536 {
                serverHost = String(value[..<colon])
                serverPort = port
            } else {
                serverHost = value
                serverPort = Self.defaultPort
            }
        }
    }

    /// nil until configured, so a caller cannot accidentally dial nowhere.
    public var serverURL: String? {
        guard let host = serverHost else { return nil }
        return "ws://\(host):\(serverPort)"
    }

    /// The single origin. Everything the server hands over as a relative path
    /// resolves against this -- persona assets, generated imagery, the house
    /// surfaces. Exactly one type constructs an origin and no port literal
    /// appears anywhere else in the source; if some asset class is not reachable
    /// through the gateway that is a server bug to file, not a second port for
    /// the client to learn.
    public var httpOrigin: String? {
        guard let host = serverHost else { return nil }
        return "http://\(host):\(serverPort)"
    }

    /// Build a URL against the single origin, or nil when no house is
    /// configured. Every HTTP call in the client goes through this.
    ///
    /// It exists so the rule can be checked by grep rather than by reading:
    /// exactly one type constructs an origin, and no port literal appears
    /// anywhere else in the source. `PersonaStore` in Valinor hardcoded :8766
    /// for persona JSON and fired one request per persona at a port nothing
    /// answered on, hanging until a 60 second timeout -- dead for two months
    /// behind a second bug in the decode. `LocalModelManager` still carries the
    /// same literal today.
    ///
    /// Returning nil rather than a URL against an empty host is the other half
    /// of the no-default-host decision: a surface that cannot build its URL
    /// reports itself unavailable, which is honest, instead of dialling nowhere
    /// and waiting out a timeout.
    public func url(_ path: String) -> URL? {
        guard let origin = httpOrigin else { return nil }
        return URL(string: origin + path)
    }

    /// A request against the house, carrying the device token.
    ///
    /// Every HTTP call in the client goes through this rather than building a
    /// URLRequest of its own, for the same reason every URL goes through
    /// `url(_:)`: so "does this carry credentials" is answered by grep instead
    /// of by auditing each call site. A site that forgets the header does not
    /// fail loudly -- it gets a 401 and reports the house unreachable, which
    /// looks exactly like a network problem.
    public func request(_ path: String, timeout: TimeInterval = 8) -> URLRequest? {
        guard let url = url(path) else { return nil }
        var request = URLRequest(url: url)
        request.timeoutInterval = timeout
        request.cachePolicy = .reloadIgnoringLocalCacheData
        authorize(&request)
        return request
    }

    /// Attach the token, if there is one. Separate so the WebSocket handshake
    /// can use it too -- URLSessionWebSocketTask takes a URLRequest, which is
    /// the only reason a header is possible there at all.
    public func authorize(_ request: inout URLRequest) {
        guard let token = deviceToken else { return }
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
    }

    /// An asset URL with the token in the QUERY STRING, for the one case that
    /// cannot send a header: `AsyncImage` takes a URL, not a URLRequest, so a
    /// card's picture and a persona's clip have no other way to authenticate.
    ///
    /// Second choice, and only used where it is the only choice -- a query
    /// string lands in server logs where a header does not. It exists because
    /// the alternative is every image in the client 401ing the moment a house
    /// is reached from off its own machine.
    public func assetURL(_ path: String) -> URL? {
        guard let base = url(path) else { return nil }
        guard let token = deviceToken else { return base }
        var components = URLComponents(url: base, resolvingAgainstBaseURL: false)
        var items = components?.queryItems ?? []
        items.append(URLQueryItem(name: "token", value: token))
        components?.queryItems = items
        return components?.url ?? base
    }

    private init() {}
}

public extension Notification.Name {
    /// The house address changed -- named, or cleared.
    ///
    /// The root view branches on `isConfigured`, and SwiftUI cannot observe a
    /// UserDefaults-backed singleton on its own. This is the redraw signal, and
    /// it fires on the clear path too: forgetting a house has to send the phone
    /// back to first run as surely as naming one takes it forward.
    static let hearthServerConfigured = Notification.Name("hearth.serverConfigured")
}

File: android-client/core/src/main/java/com/hearth/core/config/ServerConfig.kt
package com.hearth.core.config

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Request

/**
 * Where the house is, and the one type that builds an origin. Ported from the
 * iOS `ServerConfig.swift`; the parser carries across because it was paid for
 * in bugs.
 *
 * The token lives in EncryptedSharedPreferences (the Keystore-backed
 * counterpart of the iOS Keychain vault) rather than plain preferences, which
 * are readable from an unencrypted backup.
 */
class ServerConfig private constructor(context: Context) {

    private val prefs: SharedPreferences =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /**
     * Keystore-backed store for the one secret this client holds.
     *
     * Building this does real Keystore crypto, which takes long enough on a
     * cold start to be a visible stall. It must therefore never be touched
     * from the main thread: [warm] primes it on a background dispatcher and
     * [cachedToken] answers every later read from memory.
     */
    private val secure: SharedPreferences by lazy {
        val key = MasterKey.Builder(context.applicationContext)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        EncryptedSharedPreferences.create(
            context.applicationContext,
            SECURE_PREFS,
            key,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    /**
     * **There is deliberately no default host.**
     *
     * A phone supervises nothing and is never the machine running the backend,
     * so neither available default is honest: 127.0.0.1 on a phone is the
     * phone, and a LAN literal is one machine on one home network compiled
     * into a product. Unset is a distinct state rather than a missing value,
     * and the app in that state does not dial: it renders the bundled persona
     * and asks where the house is.
     */
    var serverHost: String?
        get() = prefs.getString(KEY_HOST, null)?.takeIf { it.isNotEmpty() }
        set(value) {
            val trimmed = value?.trim()
            prefs.edit().apply {
                if (!trimmed.isNullOrEmpty()) putString(KEY_HOST, trimmed) else remove(KEY_HOST)
            }.apply()
        }

    var serverPort: Int
        get() = prefs.getInt(KEY_PORT, 0).takeIf { it > 0 } ?: DEFAULT_PORT
        set(value) = prefs.edit().putInt(KEY_PORT, value).apply()

    /**
     * The device token this phone was given when it paired, or null. A house
     * exempts loopback and requires a token from everything else, so a phone
     * always needs one.
     */
    @Volatile
    private var cachedToken: String? = null

    @Volatile
    private var tokenLoaded = false

    var deviceToken: String?
        get() {
            if (tokenLoaded) return cachedToken
            // A read before warm() finished. Rare (the UI waits on warm), but
            // it must answer correctly rather than lie, so it pays the cost
            // once and caches.
            cachedToken = secure.getString(KEY_TOKEN, null)?.takeIf { it.isNotEmpty() }
            tokenLoaded = true
            return cachedToken
        }
        set(value) {
            val trimmed = value?.trim()?.takeIf { it.isNotEmpty() }
            cachedToken = trimmed
            tokenLoaded = true
            secure.edit().apply {
                if (trimmed != null) putString(KEY_TOKEN, trimmed) else remove(KEY_TOKEN)
            }.apply()
        }

    /**
     * Prime the Keystore off the main thread. Called once at startup before
     * anything reads [isPaired]; until it returns, the UI has nothing to show
     * but a splash anyway.
     */
    suspend fun warm() = withContext(Dispatchers.IO) {
        if (!tokenLoaded) {
            cachedToken = secure.getString(KEY_TOKEN, null)?.takeIf { it.isNotEmpty() }
            tokenLoaded = true
        }
    }

    /** True when someone has told this client where the house is. */
    val isConfigured: Boolean get() = serverHost != null

    /**
     * True when this phone has both an address AND a way in. Knowing where the
     * house is and being allowed through its door are two different states,
     * and a client that conflates them shows "connecting" forever instead of
     * "you need to pair".
     */
    val isPaired: Boolean get() = deviceToken != null

    /**
     * What the Connection field shows and accepts: `host` or `host:port`.
     * Clearing it clears the host rather than restoring a build-time default.
     */
    var address: String
        get() {
            val host = serverHost ?: return ""
            return if (serverPort == DEFAULT_PORT) host else "$host:$serverPort"
        }
        set(value) {
            val trimmed = value.trim()
            if (trimmed.isEmpty()) {
                serverHost = null
                serverPort = DEFAULT_PORT
                return
            }
            // Strip a pasted scheme; people paste URLs.
            var v = trimmed
            for (scheme in listOf("ws://", "wss://", "http://", "https://")) {
                if (v.startsWith(scheme)) v = v.removePrefix(scheme)
            }
            v = v.trim('/')

            // Rightmost colon only, and only when what follows is a port -- an
            // IPv6 literal is full of colons and must not be split.
            val colon = v.lastIndexOf(':')
            val port = if (colon >= 0) v.substring(colon + 1).toIntOrNull() else null
            if (colon >= 0 && port != null && port in 1..65535) {
                serverHost = v.substring(0, colon)
                serverPort = port
            } else {
                serverHost = v
                serverPort = DEFAULT_PORT
            }
        }

    /**
     * Commit a new address the way Apply and first run mean it: pointing this
     * phone at a DIFFERENT house also surrenders the old house's token.
     * Presenting house A's token to house B earns a 1008 close and nothing but
     * a reconnect loop behind it. The raw [address] setter stays token-neutral
     * on purpose, because a Test probe must not cost the pairing.
     */
    fun commitAddress(newValue: String) {
        val oldHost = serverHost
        address = newValue
        val newHost = serverHost
        if (oldHost != null && newHost != null && !newHost.equals(oldHost, ignoreCase = true)) {
            deviceToken = null
        }
    }

    /** null until configured, so a caller cannot accidentally dial nowhere. */
    val serverUrl: String? get() = serverHost?.let { "ws://$it:$serverPort" }

    /**
     * The single origin. Everything the server hands over as a relative path
     * resolves against this. Exactly one type constructs an origin and no port
     * literal appears anywhere else in the source; if some asset class is not
     * reachable through the gateway that is a server bug to file, not a second
     * port for the client to learn.
     */
    val httpOrigin: String? get() = serverHost?.let { "http://$it:$serverPort" }

    /**
     * Build a URL against the single origin, or null when no house is
     * configured. Every HTTP call in the client goes through this, so the rule
     * can be checked by grep rather than by reading.
     */
    fun url(path: String): String? = httpOrigin?.let { it + path }

    /**
     * A request against the house, carrying the device token. Every HTTP call
     * goes through this rather than building its own, so "does this carry
     * credentials" is answered by grep instead of by auditing each call site.
     */
    fun request(path: String): Request? {
        val url = url(path) ?: return null
        val builder = Request.Builder().url(url)
        authorize(builder)
        return builder.build()
    }

    /**
     * Attach the token, if there is one. Separate so the WebSocket handshake
     * can use it too.
     */
    fun authorize(builder: Request.Builder) {
        val token = deviceToken ?: return
        builder.header("Authorization", "Bearer $token")
    }

    /**
     * An asset URL with the token in the QUERY STRING, for the one case that
     * cannot send a header. On Android the image loader takes an interceptor,
     * so this is only for a surface that genuinely holds a bare URL; prefer
     * [request] everywhere else. A query string lands in server logs where a
     * header does not.
     */
    fun assetUrl(path: String): String? {
        val base = url(path) ?: return null
        val token = deviceToken ?: return base
        val sep = if (base.contains('?')) "&" else "?"
        return "$base${sep}token=$token"
    }

    companion object {
        private const val PREFS = "hearth.config"
        private const val SECURE_PREFS = "hearth.secure"
        private const val KEY_HOST = "hearth.serverHost"
        private const val KEY_PORT = "hearth.serverPort"
        private const val KEY_TOKEN = "hearth.deviceToken"

        /**
         * 18700 is Hearth's client gateway, and the port is the whole default.
         *
         * NOT 8700. The development machine runs Valinor's stack there, so a
         * Hearth build defaulting to 8700 does not fail -- it connects, and
         * comes up wearing someone else's memory, journal and personas. That
         * is a first run which looks flawless and is the worst outcome
         * available.
         */
        const val DEFAULT_PORT = 18700

        @Volatile
        private var instance: ServerConfig? = null

        fun get(context: Context): ServerConfig =
            instance ?: synchronized(this) {
                instance ?: ServerConfig(context).also { instance = it }
            }
    }
}
